src/vsr/message_header.zig

```diff
@@ -0,0 +1,1430 @@
+    size: u32 = @sizeOf(Header),
+    epoch: u32 = 0,
+    view: u32,
+    version: u16 = vsr.Version,
+    command: Command,
+    replica: u8,
+    reserved_frame: [12]u8 = [_]u8{0} ** 12,
+
+    /// Set to zero for a new view, and to a nonce from an RSV when responding to the RSV.
+    nonce: u128,
+    op: u64,
+    /// Set to `commit_min`/`commit_max` (they are the same).
+    commit: u64,
+    /// The replica's `op_checkpoint`.
+    checkpoint_op: u64,
+    reserved: [136]u8 = [_]u8{0} ** 136,
+
+    pub fn frame(header: *@This()) *Header {
+        return std.mem.bytesAsValue(Header, std.mem.asBytes(header));
+    }
+
+    pub fn frame_const(header: *const @This()) *const Header {
+        return std.mem.bytesAsValue(Header, std.mem.asBytes(header));
+    }
+
+    fn invalid(self: *const @This()) ?[]const u8 {
+        assert(self.command == .start_view);
+        if (!stdx.zeroed(&self.reserved)) return "reserved != 0";
+        return null;
+    }
+};
+
+const RequestStartView = extern struct {
+    checksum: u128 = 0,
+    checksum_body: u128 = 0,
+    reserved_nonce: u128 = 0,
+    cluster: u32,
+    size: u32 = @sizeOf(Header),
+    epoch: u32 = 0,
+    view: u32,
+    version: u16 = vsr.Version,
+    command: Command,
+    replica: u8,
+    reserved_frame: [12]u8 = [_]u8{0} ** 12,
+
+    nonce: u128,
+    reserved: [160]u8 = [_]u8{0} ** 160,
+
+    pub fn frame(header: *@This()) *Header {
+        return std.mem.bytesAsValue(Header, std.mem.asBytes(header));
+    }
+
+    pub fn frame_const(header: *const @This()) *const Header {
+        return std.mem.bytesAsValue(Header, std.mem.asBytes(header));
+    }
+
+    fn invalid(self: *const @This()) ?[]const u8 {
+        assert(self.command == .request_start_view);
+        if (self.size != @sizeOf(Header)) return "size != @sizeOf(Header)";
+        if (self.checksum_body != checksum_body_empty) return "checksum_body != expected";
+        if (!stdx.zeroed(&self.reserved)) return "reserved != 0";
+        return null;
+    }
+};
+
+const RequestHeaders = extern struct {
+    checksum: u128 = 0,
+    checksum_body: u128 = 0,
+    reserved_nonce: u128 = 0,
+    cluster: u32,
+    size: u32 = @sizeOf(Header),
+    epoch: u32 = 0,
+    view: u32,
+    version: u16 = vsr.Version,
+    command: Command,
+    replica: u8,
+    reserved_frame: [12]u8 = [_]u8{0} ** 12,
+
+    /// The minimum op requested (inclusive).
+    op_min: u64,
+    /// The maximum op requested (inclusive).
+    op_max: u64,
+    reserved: [160]u8 = [_]u8{0} ** 160,
+
+    pub fn frame(header: *@This()) *Header {
+        return std.mem.bytesAsValue(Header, std.mem.asBytes(header));
+    }
+
+    pub fn frame_const(header: *const @This()) *const Header {
+        return std.mem.bytesAsValue(Header, std.mem.asBytes(header));
+    }
+
+    fn invalid(self: *const @This()) ?[]const u8 {
+        assert(self.command == .request_headers);
+        if (self.size != @sizeOf(Header)) return "size != @sizeOf(Header)";
```

Comment:
makes me wonder if we should check that the view is zero here and for RequestPrepare (and for RequestReply)? I _think_ we are not actually using RequestHeader's view for anything?